design/007-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,421 @@
+whether the results are still in sync with the data using `renku status`. In
+this case, Renku checks the object hashes that were recorded previously against
+the data currently found in the remote location (this might take a while over
+the network - some data sources might provide pre-computed hashes). If the
+hashes differ, Renku offers to update the results.
+
+#### 3. Data storage for research project
+
+Willow is interested in a large, protected dataset that is offered for free to
+researchers by an external organization. She has to accept a data access
+agreement to access the data, at which points she receives credentials to the
+bucket. To explore the dataset, she creates a Renku project and mounts the S3
+bucket using the credentials.
+
+After some exploration of the dataset, Willow decides to start a research
+project with the dataset. She makes a copy of the dataset in her own private S3
+bucket for two reasons: (1) to make sure she maintains access to the data
+independent of the external organization, and (2) because she prefers to keep
+the project’s raw, intermediate, and final data products in the same place. She
+creates a Renku project for her work and mounts the bucket with her credentials.
+Throughout the research project, she reads the raw data from the mounted
+directory and saves intermediate data files and figures in separate folders in
+the same S3 bucket.
+
+#### 4. Access remote data but make local copies
+
+Sharing data in remote S3 buckets might not be only for large data that cannot
+be stored locally; in some cases it might be prudent to make local copies. Users
+frequently write their own supporting code for handling this.
+
+##### In the case where the data is small:
+
+1. write a script to fetch the data and store it in a subfolder of the project
+2. add the script to git
+3. add the subfolder to gitignore
+
+This way, the data is local and can be worked on in the context of the project.
+
+##### In the case where the data is large:
+
+1. write a script to fetch the data and store it on an external drive or another filesystem
+2. add the script to git
+3. write scripts that reference the external drive/filesystem and reduce the data to a manageable size
+4. add those scripts to git
+5. if the reduced data is very small, I will add it to git as well, if it is too large, I will put it in a gitignored subfolder
+
+In this case, if the project is cloned to a new computer or the external
+drive/filesystem is not available, the data can be fetched and reduced again
+easily.
+
+
+### Use-cases
+
+From the user stories, two primary high-level use-cases emerge:
+
+1. use data from remote storage as-is by remote-mounting it and remembering the
+   source in renku metadata. This should support reading and writing.
+
+2. use external storage as a source/sink but make local copies to speed up
+   access for iterative calculations. In principle, this should also support
+   reading and writing.
```

Comment:
So we need some kind of a system/method for avoiding conflicts when local data is written to, right? If two people mount a large dataset and download the data locally around the same time what is the final version that is in the S3 bucket at the end? When are changes reflected in the S3 bucket when a user makes changes locally? For example, do we sync immediately or do we depend on the user for syncing or do we sync at specific times (i.e. when the user stops their session)?

Is this something we should discuss here in the RFC or not?

Reply:
Hmm that's a good question. I'm not sure this needs to be our responsibility tbh. If I think of how it works on other shared infrastructures, the infrastructure provider isn't making sure that users don't overwrite each other's stuff in shared directories.